Allow function pointer arguments to be equal

xbauch opened the pull request:
for the function-call goto-harness.

We add a new command-line option `treat-pointers-equal` that takes a list of
formal parameter names and results in a harness that only generates constructor
for one of the pointers. The other pointers are simply assigned with the
constructed one.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

petr-bauch committed:
Allow function pointer arguments to be equal

for the function-call goto-harness.

We add a new command-line option `treat-pointers-equal` that takes a list of
formal parameter names and results in a harness that only generates constructor
for one of the pointers. The other pointers are simply assigned with the
constructed one.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9db8bd4aef5c103c755cedabff4cbf9746bac26f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/134853929

NlightNFotis reviewed (approved):
LGTM

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -763,3 +783,17 @@ code_blockt recursive_initializationt::build_dynamic_array_constructor(
+
+bool recursive_initializationt::needs_freeing(const exprt &expr) const
+{
+  if(expr.type().id() != ID_pointer)
```
Function pointers also shouldn't be freed but we're already fixing this in a follow up PR

hannes-steffenhagen-diffblue reviewed (commented):
This is currently only capable of expressing that a single set of function params should be treated equal, I'd like this to be able to specify multiple (non-overlapping I'd hope?) sets of possibly equal pointers.

Also, these should be _maybe_ equal, but as currently specified they'll _definitely_ be equal, which is a bit unfortunate (as now we won't be covering the case where they're not equal, we'd have to generate a different harness for these again).

petr-bauch committed:
Extend to clusters of equal parameters

Specified by a semi-colon separated list of comma separated names.

xbauch commented:
@hannes-steffenhagen-diffblue both sets of equal pointers and conditional equality should be implemented now.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: dcf0ea3a5789ebcf3b8214002f0f71ec621f2662).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/136572662

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -112,6 +115,21 @@ void function_call_harness_generatort::handle_option(
+  {
+    for(auto const &value : values)
+    {
+      for(auto const &param_cluster : split_string(value, ';'))
```
Super minor nitpick that `;` might not necessarily be the best separator to use here because on most shells it’s also a terminator, but tbh I don’t think it’s big enough of a problem to worry about here.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -763,3 +817,52 @@ code_blockt recursive_initializationt::build_dynamic_array_constructor(
+
+bool recursive_initializationt::needs_freeing(const exprt &expr) const
+{
+  if(expr.type().id() != ID_pointer || expr.type().id() == ID_code)
```
This could just be `return expr.type().id() = ID_pointer && expr.type().id() != ID_code;`

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -280,6 +321,19 @@ bool recursive_initializationt::should_be_treated_as_array(
          initialization_config.pointers_to_treat_as_arrays.end();
 }
 
+optionalt<std::size_t>
```
Small preference to have a (strong?) typedef for this `size_t` so we know that it refers to an “equal cluster”

xbauch replied:
Done.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -763,3 +817,52 @@ code_blockt recursive_initializationt::build_dynamic_array_constructor(
+    initialization_config.arguments_may_be_equal)
+  {
+    // in equality cluster but not common origin -> free if equal to origin
+    const auto condition =
```
I don’t quite understand the logic, shouldn’t it be the other way round (i.e. don’t free if not common origin, unless your value is different than origin)?

xbauch replied:
Off by negation: good catch. Fixed.

petr-bauch committed:
Extend to allow conditional equality

between pointer arguments. Whether or not they will be equal is now a
non-deterministic choice. Freeing pointer had to be extended.
